sjsonnet/src/sjsonnet/Util.scala

```diff
@@ -72,26 +72,115 @@ object Util {
+    val safeEnd = math.max(safeStart, math.min(endCodePointOffset, unicodeLength))
+
+    if (safeStart == safeEnd) {
+      val utf16Pos = if (safeStart == 0) 0 else s.offsetByCodePoints(0, safeStart)
+      (utf16Pos, utf16Pos)
+    } else {
+      val startUtf16 = if (safeStart == 0) 0 else s.offsetByCodePoints(0, safeStart)
+      val endUtf16 = s.offsetByCodePoints(startUtf16, safeEnd - safeStart)
+      (startUtf16, endUtf16)
+    }
   }
-  def sliceStr(s: String, start: Int, end: Int, step: Int): String = {
-    if (start >= end || start >= s.length) {
+
+  private def sliceStr(s: String, start: Int, end: Int, step: Int): String = {
+    val unicodeLength = s.codePointCount(0, s.length)
+    if (start >= end || start >= unicodeLength) {
       ""
-    } else
+    } else {
       step match {
-        case 1 => s.slice(start, end)
+        case 1 =>
+          val (startUtf16, endUtf16) = codePointOffsetsToStringIndices(s, start, end)
+          s.substring(startUtf16, endUtf16)
         case _ =>
-          val range = start until end by step
-          new String(range.dropWhile(_ < 0).takeWhile(_ < s.length).map(s).toArray)
+          val result = new java.lang.StringBuilder()
+          var sIdx = 0
+          var codepointIndex = 0
+
+          // Skip to start codepoint position
+          while (sIdx < s.length && codepointIndex < start) {
+            val cp = s.codePointAt(sIdx)
+            sIdx += Character.charCount(cp)
+            codepointIndex += 1
+          }
+
+          // Collect every `step`th codepoint until `end`
+          var rel = 0 // relative index from start
+          while (sIdx < s.length && codepointIndex < end) {
+            val c = s.charAt(sIdx)
+            if (Character.isSurrogate(c)) {
+              // Handle surrogate pair
+              val cp = s.codePointAt(sIdx)
+              if (rel % step == 0) {
+                result.append(Character.toString(cp))
+              }
+              sIdx += Character.charCount(cp)
+            } else {
+              // Single char, non-surrogate
+              if (rel % step == 0) {
+                result.append(c)
+              }
+              sIdx += 1
+            }
+            codepointIndex += 1
+            rel += 1
+          }
+          result.toString
       }
+    }
   }
-  def sliceStr(s: String, start: Option[Int], end: Option[Int], step: Option[Int]): String = {
-    sliceStr(s, start.getOrElse(0), end.getOrElse(s.length), step.getOrElse(1))
+
+  /**
+   * Compares two strings by Unicode codepoint values rather than UTF-16 code units.
+   * This ensures that strings with characters above U+FFFF (which require surrogate pairs
+   * in UTF-16) are compared correctly according to their Unicode codepoint values.
+   */
+  def compareStringsByCodepoint(s1: String, s2: String): Int = {
+    val n1 = s1.length
+    val n2 = s2.length
+    var i1 = 0
+    var i2 = 0
+    while (i1 < n1 && i2 < n2) {
```

Comment:
On reflection, I think that the character-by-character approach will be inefficient compared to [Arrays.mismatch](https://docs.oracle.com/javase/9/docs/api/java/util/Arrays.html#mismatch-char:A-char:A-) for seeking to the first unmatched position then comparing the codepoints at those indices.

That approach would clearly be correct in the case where we're comparing BMP and non-BMP characters. 

But what if we have two strings where the first mismatched characters are non-BMP characters that have the same high surrogate? I think that case is fine, too: we'd be comparing the low surrogates to each other, but they correspond to the lower bits of the final codepoints and we already know that they have the same upper bits.

Let me take a stab at this optimization.